### PR TITLE
dump: Better printing of rec groups and sub types

### LINKED ISSUE #1359: Dump individual types within rec groups, not whole rec group at once
`wasm-tools dump my-module-that-uses-rec-groups.wasm` is not super useful at the moment.

### PR DESCRIPTION
**Pull request by fitzgen**
Still not perfect, but good enough for now. Follow up improvements always welcome.

Fixes #1359

### CODE REVIEW

File: crates/wasmparser/src/readers/core/types.rs [hunk 1]
```diff
@@ -306,65 +306,102 @@ pub struct RecGroup {
...
             RecGroupInner::Explicit(types) => types,
-        }
+        };
+        types.iter_mut().map(|(_, ty)| ty)
     }
 
     /// Returns an owning iterator of all subtypes in this recursion
     /// group.
     pub fn into_types(self) -> impl ExactSizeIterator<Item = SubType> {
         return match self.inner {
```
Comment (alexcrichton) [inline crates/wasmparser/src/readers/core/types.rs:356]: Could this call `into_types_and_offsets` and use `map` to discard the offset?

**Review by alexcrichton — COMMENTED** [review]
Could you add a test to `test/cli/dump` which exhibits the new printing? It looks a bit overly verbose to me so I'm curious how it fares with rec groups that have multiple types

File: tests/cli/dump/alias.wat.stdout [hunk 2]
```diff
@@ -30,7 +30,17 @@
         | 01 00 00 00
    0x57 | 01 04       | type section
    0x59 | 01          | 1 count
-   0x5a | 60 00 00    | [type 0] RecGroup { inner: Implicit(SubType { is_final: true, supertype_idx: None, composite_type: Func(FuncType { params: [], results: [] }) }) }
+--- rec group 0 (implicit) ---
+   0x5a | 60 00 00    | [type 0] SubType {
+    is_final: true,
```
[the remark this answers is not in the document]
Comment (fitzgen) [reply tests/cli/dump/alias.wat.stdout:35]: I don't think `std::fmt` provides a way to do that does it?

[the remark this answers is not in the document]
Comment (fitzgen) [reply tests/cli/dump/alias.wat.stdout:35]: I can put it all back on one line, if you prefer, but it makes structs super hard to read.

[the remark this answers is not in the document]
Comment (fitzgen) [reply tests/cli/dump/alias.wat.stdout]: Put it back on one line.

**Review by alexcrichton — APPROVED** [review]

File: tests/cli/dump/alias.wat.stdout (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexcrichton) [reply tests/cli/dump/alias.wat.stdout]: Oh multiline I think is fine but I was hoping we could split the output by lines and print each line separately with padding at the front. If you want to change this to print multiline I think that's ok too though as-is was before (I went ahead and merged as I figured this'd be a good follow-up)

**Pull request merged**